pfb/base.py

```diff
@@ -0,0 +1,159 @@
+            symbols = []
+            for symbol in t["symbols"]:
+                symbols.append(method(symbol))
+            t["symbols"] = symbols
+    default = field.get("default")
+    if is_enum_ and default:
+        field["default"] = method(default)
+
+
+def avro_record(node_id, node_name, values, relations):
+    node = {"id": node_id, "name": node_name, "object": values, "relations": relations}
+    return node
+
+
+class PFBBase(object):
+    open_mode = "r"
+
+    def __init__(self, file_or_path):
+        self._file_or_path = file_or_path
+        self._is_base64 = None
+        self._schema = None
+        self._metadata = None
+
+    def __enter__(self):
+        self._file_obj = (
+            self._file_or_path
+            if hasattr(self._file_or_path, "read")
+            else open(self._file_or_path, self.open_mode)
+        )
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._file_obj.close()
+
+    @property
+    def name(self):
+        return self._file_obj.name
+
+    @property
+    def isatty(self):
+        return self._file_obj.isatty()
+
+    @property
+    def schema(self):
+        return self._schema
+
+    @property
+    def metadata(self):
+        return self._metadata
+
+    def set_schema(self, schema):
+        self._schema = schema
+
+    def set_metadata(self, metadata):
+        self._metadata = metadata
+
+    def prepare_base64_cache(self):
+        if self._is_base64 is None:
+            self._is_base64 = {}
+            for node in self._schema:
+                self._is_base64[node["name"]] = fields = {}
+                for field in node["fields"]:
+                    fields[field["name"]] = is_enum(field["type"])
+
+    def is_base64(self, node_name, field_name):
+        self.prepare_base64_cache()
+        return self._is_base64[node_name].get(field_name, False)
+
+    def make_empty_record(self, node_name):
+        fields = {}
+        for node in self.schema:
+            if node["name"] == node_name:
+                for field in node["fields"]:
+                    type_ = field["type"]
+                    stack = deque([field["type"]])
+                    while stack:
+                        t = stack.pop()
+                        if isinstance(t, list):
+                            stack.extend(t)
+                        elif t == "long":
+                            type_ = "long"
+                            break
+                        elif t == "float":
+                            type_ = "float"
+                            break
+                        elif isinstance(t, dict) and t["type"] == "enum":
+                            type_ = "enum"
+                            break
+
+                    fields[field["name"]] = type_
+
+        uid = uuid.uuid4()
+
+        val = {}
+        for x in fields:
+            if fields[x] in ("long", "float"):
+                val[x] = 0
+            elif fields[x] == "enum":
```

Comment:
Did you test this? In Avro spec `None` can't be used for enums and in dictionary it can be either mandatory type (then it would be enum and `None` is not allowed) or optional type with `null` in union (in this case, it will be `None` only for `null` type to be first in union: [Avro spec](https://avro.apache.org/docs/current/spec.html#schema_complex)).

Reply:
Oh yes, it was because `""` is not a working value too. But as you pointed out, it is not quite proper now either. I'll try to make this more reasonable.